nagios.go

```diff
@@ -1,55 +1,28 @@
 package main
 
 import (
-	"fmt"
 	"os/exec"
+	"path/filepath"
 )
 
-// GetNagiosStatusData is a task factory for tasks that fetch a JSON representation of
-// the provided project. The task uses outFor and errOutFor to get io.Writers
-// to write, respectively, the JSON output and any eventual error message.
-func GetNagiosStatusData(project, pod string, outFor, errOutFor projectResourceWriterCloserFactory) Task {
-	return retrieveNagiosData(func() *exec.Cmd {
-		return exec.Command("oc", "exec", pod, "--", "cat", "/var/log/nagios/status.dat")
-	}, project, pod, "status", outFor, errOutFor)
-}
-
-// GetNagiosHistoricalData is a task factory for tasks that fetch a JSON representation of
-// the provided project. The task uses outFor and errOutFor to get io.Writers
-// to write, respectively, the JSON output and any eventual error message.
-func GetNagiosHistoricalData(project, pod string, outFor, errOutFor projectResourceWriterCloserFactory) Task {
-	return retrieveNagiosData(func() *exec.Cmd {
-		return exec.Command("oc", "exec", pod, "--", "tar", "-c", "-C", "/var/log/nagios", "archives")
-	}, project, pod, "history", outFor, errOutFor)
+// GetNagiosStatusData is a task factory for tasks that fetch Nagios status from
+// the given pod in project.
+func GetNagiosStatusData(r Runner, project, pod string) Task {
+	return func() error {
+		cmd := exec.Command("oc", "exec", pod, "--", "cat", "/var/log/nagios/status.dat")
+		fname := pod + "_status.dat"
+		path := filepath.Join("projects", project, "nagios", fname)
+		return r.Run(cmd, path)
+	}
 }
 
-// A getNagiosDataCmdFactory generates commands to get the raw nagios status data
-// in the provided project from the provided pod
-type getNagiosDataCmdFactory func() *exec.Cmd
-
-// retrieveNagiosData will dump a JSON representation of the current nagios status into
-// the io.Writer returned from outFor and log any errors it encounters doing so
-// into the io.Writer returned from errOutFor.
-func retrieveNagiosData(cmdFactory getNagiosDataCmdFactory, project, pod, nagiosResourceType string, outFor, errOutFor projectResourceWriterCloserFactory) Task {
+// GetNagiosHistoricalData is a task factory for tasks that fetch Nagios
+// archives from the given pod in project.
+func GetNagiosHistoricalData(r Runner, project, pod string) Task {
 	return func() error {
-		stdout, stdoutCloser, err := outFor(project, pod+"-"+nagiosResourceType)
-		if err != nil {
-			return err
-		}
-		defer stdoutCloser.Close()
-
-		stderr, stderrCloser, err := errOutFor(project, pod+"-"+nagiosResourceType)
-		if err != nil {
-			return err
-		}
-		defer stderrCloser.Close()
-
-		cmd := cmdFactory()
-
-		if err := runCmdCaptureOutput(cmd, stdout, stderr); err != nil {
-			fmt.Fprint(stderr, err)
-			return err
-		}
-		return nil
+		cmd := exec.Command("oc", "exec", pod, "--", "tar", "-c", "-C", "/var/log/nagios", "archives")
+		fname := pod + "_history.tar"
+		path := filepath.Join("projects", project, "nagios", fname)
```

Comment:
And finally here the path should not have changed.

Now paths are fully constructed in a single place, which I consider better than before. However there's still the problem I mentioned on IRC, manually writing these paths is error prone, we could get into a situation that a file name clashes with an existing directory or vice-versa.